lib/kubernetes/client.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+class Kubernetes::Client
+  def initialize(kubernetes_cluster, session)
+    @session = session
+    @kubernetes_cluster = kubernetes_cluster
+    @load_balancer = LoadBalancer.where(name: kubernetes_cluster.services_load_balancer_name).first
+  end
+
+  def is_service_deleted(svc)
+    !!svc.dig("metadata", "deletionTimestamp")
+  end
+
+  # Returning an array in this format:
+  # [[src_port_0, dst_port_0], [src_port_1, dst_port_1],...]
+  def lb_desired_ports(svc)
+    svc.dig("spec", "ports").map { |port| [port["port"], port["nodePort"]] }
+  end
+
+  def load_balancer_hostname_missing?(svc)
+    svc.dig("status", "loadBalancer", "ingress", "hostname").to_s.empty?
+  end
+
+  def kubectl(cmd)
+    @session.exec!("sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf #{cmd}")
+  end
+
+  def set_load_balancer_host_name(svc, hostname)
+    patch_data = JSON.generate({
+      "status" => {
+        "loadBalancer" => {
+          "ingress" => [{"hostname" => hostname}]
+        }
+      }
+    })
+    kubectl("-n #{svc.dig("metadata", "namespace")} patch service #{svc.dig("metadata", "name")} --type=merge -p '#{patch_data}' --subresource=status")
+  end
+
+  def sync_kubernetes_services
+    k8s_svc_raw = kubectl("get service --all-namespaces --field-selector spec.type=LoadBalancer -ojson")
+    JSON.parse(k8s_svc_raw)["items"].each do |svc|
+      reconcile_kubernetes_service(svc)
+    end
+  end
+
+  def reconcile_kubernetes_service(svc)
+    return if is_service_deleted(svc)
+
+    desired_ports = lb_desired_ports(svc)
+    if @load_balancer.nil?
+      raise "services LoadBalancer does not exist."
+    end
+
+    extra_vms, missing_vms = vm_diff_for_lb
+    missing_vms.each { |missing_vm| @load_balancer.add_vm(missing_vm) }
+    extra_vms.each { |extra_vm| @load_balancer.detach_vm(extra_vm) }
+
+    extra_ports, missing_ports = port_diff_for_lb(desired_ports)
+    extra_ports.each { |port| @load_balancer.remove_port(port) }
+    missing_ports.each { |port| @load_balancer.add_port(port[0], port[1]) }
+
+    set_load_balancer_host_name(svc, @load_balancer.hostname) if @load_balancer.strand.label == "wait"
+  end
+
+  def vm_diff_for_lb
```

Comment:
this function shouldn't be in the client, as it has nothing to do with Kubernetes itself. We should make sure to handle this in LB control functions.